Tests/UnitTests/Sources/GREYRunLoopSpinnerTest.m

```diff
@@ -0,0 +1,317 @@
+                           @"drains after the condition is met. However, the counter source may"
+                           @"have been handled once after the condition was met.");
+
+  XCTAssertTrue(result,
+                 @"Spin result should be YES. The condition was met.");
+}
+
+- (void)testSpinningEmptyMode {
+  GREYRunLoopSpinner *spinner = [[GREYRunLoopSpinner alloc] init];
+
+  spinner.maxSleepInterval = 0;
+  spinner.minRunLoopDrains = 2;
+  spinner.timeout = DBL_MAX;
+
+  __block int conditionCheckCountLimit = 20;
+  __block int currentConditionCheckCount = 0;
+
+  [self changeActiveModeToSpinnerTestMode];
+
+  BOOL result = [spinner spinWithStopConditionBlock:^BOOL{
+    currentConditionCheckCount++;
+    return currentConditionCheckCount > conditionCheckCountLimit;
+  }];
+
+  XCTAssertTrue(result,
+                @"Spin result should be YES. The condition was eventually met.");
+}
+
+- (void)testChecksConditionAtLeastOncePerDrain {
+  GREYRunLoopSpinner *spinner = [[GREYRunLoopSpinner alloc] init];
+
+  spinner.maxSleepInterval = 0;
+  spinner.minRunLoopDrains = 0;
+  spinner.timeout = DBL_MAX;
+
+  __block int conditionCheckCountLimit = 20;
+  __block int currentConditionCheckCount = 0;
+  __block int lastConditionCheckDrainCount = gDrainCountForSpinnerTest;
+
+  [self changeActiveModeToSpinnerTestMode];
+  [self setupDrainCounter];
+
+  BOOL result = [spinner spinWithStopConditionBlock:^BOOL{
+    XCTAssertLessThanOrEqual(gDrainCountForSpinnerTest, lastConditionCheckDrainCount + 1,
+                             @"The drain count should not increase by more than 1 per stop"
+                             @"condition check. The spinner should check this condition at least"
+                             @"once every drain.");
+    lastConditionCheckDrainCount = gDrainCountForSpinnerTest;
+    currentConditionCheckCount++;
+    return currentConditionCheckCount > conditionCheckCountLimit;
+  }];
+
+  XCTAssertTrue(result,
+                @"Spin result should be YES. The condition was eventually met.");
+  XCTAssertLessThanOrEqual(gDrainCountForSpinnerTest, lastConditionCheckDrainCount + 1,
+                @"Spinner should not iniate any new drains after the condition was met.");
+}
+
+// Note, though it was shown to be robust when it was added, this test is very likely to become
+// flaky. If it does, remove it.
+- (void)testTimeout {
+  GREYRunLoopSpinner *spinner = [[GREYRunLoopSpinner alloc] init];
+
+  CFTimeInterval timeout = 0.2;
+  CFTimeInterval timeoutTime = CACurrentMediaTime() + timeout;
+
+  spinner.maxSleepInterval = 0;
+  spinner.minRunLoopDrains = 0;
+  spinner.timeout = timeout;
+
+
+  [self changeActiveModeToSpinnerTestMode];
+
+  BOOL result = [spinner spinWithStopConditionBlock:^BOOL{
+    return NO;
+  }];
+
+  XCTAssertFalse(result,
+                 @"Spin result should be NO. The condition was never met.");
+  XCTAssertEqualWithAccuracy(timeoutTime, CACurrentMediaTime(), .0001,
```

Comment:
.xxx => 0.xxx

Also this is flaky, change 0.0001 to 0.01 or something because machines on Travis are incredibly slow.